src/signerkey.js

```diff
@@ -0,0 +1,68 @@
+import xdr from './generated/stellar-xdr_generated';
+import { StrKey, encodeCheck, decodeCheck } from './strkey';
+
+export class SignerKey {
+  static decodeAddress(address) {
+    const signerKeyMap = {
+      ed25519PublicKey: xdr.SignerKey.signerKeyTypeEd25519,
+      preAuthTx: xdr.SignerKey.signerKeyTypePreAuthTx,
+      sha256Hash: xdr.SignerKey.signerKeyTypeHashX,
+      signedPayload: xdr.SignerKey.signerKeyTypeEd25519SignedPayload
+    };
+
+    const vb = StrKey.getVersionByteForPrefix(address);
+    const encoder = signerKeyMap[vb];
+    if (!encoder) {
+      throw new Error(`invalid signer key type (${vb})`);
+    }
+
+    const raw = decodeCheck(vb, address);
+    switch (vb) {
+      case 'signedPayload':
+        return encoder(
+          new xdr.SignerKeyEd25519SignedPayload({
+            ed25519: raw.slice(0, 32),
+            payload: raw.slice(32 + 4)
+          })
+        );
+
+      case 'ed25519PublicKey': // falls through
+      case 'preAuthTx': // falls through
+      case 'sha256Hash': // falls through
+      default:
+        return encoder(raw);
+    }
+  }
+
+  static encodeSignerKey(signerKey) {
```

Comment:
comment with description and types here as well?

Reply:
Yep https://github.com/stellar/js-stellar-base/pull/520/commits/342a98758cee7584217107db02ece7624a9d4f54